### PR TITLE
Add `sec_to_time` function to OpenSearch

### PR DESCRIPTION
**Pull request by GabeFernandez310**
### Description
Adds the sec_to_time function to the OpenSearch SQL plugin according to [MySQL documentation](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_sec-to-time).

 
### Issues Resolved
https://github.com/opensearch-project/sql/issues/722
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass, including unit test, integration test and doctest
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
  - [X] New functionality has user manual doc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 1]
```diff
@@ -1364,6 +1373,38 @@ private ExprValue exprQuarter(ExprValue date) {
...
+        return new ExprTimeValue(String.format("%02d:%02d:%02d", hours, minutes, seconds));
+      }
+
+      return new ExprTimeValue(String.format("%02d:%02d:%02d.%s",
+          hours,
+          minutes,
+          seconds,
+          fraction.substring(2)));
+
+    } catch (Exception e) {
```
Comment (acarbonetto) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1402]: Please don't do this.  This will catch any error and hide it. 
Whenever using try...catch, be as specific as possible with the Exception and the line numbers (so if you're trying to catch parsing failures, include the try...catch around your parsing statements only.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 2]
```diff
@@ -1173,6 +1176,50 @@ public void quarter() {
     assertEquals(integerValue(4), eval(expression));
   }
 
+  private static Stream<Arguments> getTestDataForSecToTime() {
+    return Stream.of(
+        Arguments.of(1, "00:00:01"),
+        Arguments.of(2378, "00:39:38"),
+        Arguments.of(6897, "01:54:57")
+        //TODO: Discuss whether negative values should be supported like in MySQL.
```
Comment (acarbonetto) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1184]: let's note this in the documentation that it's a limitation for the function

File: docs/user/dql/functions.rst [hunk 3]
```diff
@@ -2096,6 +2096,28 @@ Example::
     | 3                             |
     +-------------------------------+
 
+SEC_TO_TIME
+-----------
+
+Description
+>>>>>>>>>>>
+
+Usage: sec_to_time(integer) returns the time in HH:mm:ss format.
```
Comment (acarbonetto) [inline docs/user/dql/functions.rst:2105]: don't forget the nanoseconds in the output

**Review by acarbonetto — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 4]
```diff
@@ -1364,6 +1373,38 @@ private ExprValue exprQuarter(ExprValue date) {
...
+    if (val < 0) {
+      return Math.ceil(val);
+    }
+    return Math.floor(val);
+  }
+
+  private ExprValue exprSecToTime(ExprValue totalSeconds) {
+
+    try {
+      int hours = totalSeconds.integerValue() / 3600;
```
Comment (GabeFernandez310) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1386]: Use `LocalTime.MIN + Duration.ofSeconds(arg)`

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java (hunk 2 above)
Comment (GabeFernandez310) [reply core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java]: Fixed in https://github.com/Bit-Quill/opensearch-project-sql/pull/219/commits/9ae0057b2cc009e94c396c8713bc68d57020b1db

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 1 above)
Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java]: Fixed in https://github.com/Bit-Quill/opensearch-project-sql/pull/219/commits/9ae0057b2cc009e94c396c8713bc68d57020b1db

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 5]
```diff
@@ -1364,6 +1375,34 @@ private ExprValue exprQuarter(ExprValue date) {
...
+  private ExprValue exprSecToTime(ExprValue totalSeconds) {
+    return new ExprTimeValue(LocalTime.MIN.plus(Duration.ofSeconds(totalSeconds.longValue())));
+  }
+
+  private long formatNanos(double nanos) {
+    while (nanos < 99999999) {
+      nanos *= 10;
+    }
+    return (long)nanos;
+  }
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1392]: This would convert 42 millis and 42 nanos to the same number.
Test with 3.14 and 1.000000042

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 6]
```diff
@@ -1173,6 +1176,47 @@ public void quarter() {
...
+    lenient().when(missingRef.valueOf(env)).thenReturn(missingValue());
+    FunctionExpression expr = DSL.sec_to_time(
+        DSL.literal(new ExprIntegerValue(seconds)));
+
+    assertEquals(TIME, expr.type());
+    assertEquals(new ExprTimeValue(expected), eval(expr));
+  }
+
+  @Test
+  public void testSecToTimeWithDecimal() {
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1200]: Try different values in this test to prove you convert works fine.

File: docs/user/dql/functions.rst [hunk 7]
```diff
@@ -2096,6 +2096,45 @@ Example::
     | 3                             |
     +-------------------------------+
 
+SEC_TO_TIME
+-----------
+
+Description
+>>>>>>>>>>>
+
+Usage: sec_to_time(number) returns the time in HH:mm:ss format.
```
Comment (Yury-Fridlyand) [inline docs/user/dql/functions.rst:2105]: Mention that function round around 24 clock. Value equal to 25x3600 produce the same result as 3600.

File: core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java [hunk 8]
```diff
@@ -1173,6 +1176,47 @@ public void quarter() {
     assertEquals(integerValue(4), eval(expression));
   }
 
+  private static Stream<Arguments> getTestDataForSecToTime() {
+    return Stream.of(
+        Arguments.of(1, "00:00:01"),
+        Arguments.of(2378, "00:39:38"),
+        Arguments.of(6897, "01:54:57")
```
Comment (Yury-Fridlyand) [inline core/src/test/java/org/opensearch/sql/expression/datetime/DateTimeFunctionTest.java:1183]: Add a test for negative values and values > 24x3600

**Review by GumpacG — APPROVED** [review]

**Review by Yury-Fridlyand — APPROVED** [review]

**Review by matthewryanwells — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 9]
```diff
@@ -1364,6 +1376,46 @@ private ExprValue exprQuarter(ExprValue date) {
...
+   * @param seconds and ExprDoubleValue or ExprFloatValue for the seconds
+   * @return A LONG representing the nanoseconds portion
+   */
+  private long formatNanos(ExprValue seconds) {
+    String nanoStringVal = seconds.doubleValue().toString();
+
+    //Obtains only the decimal part of the seconds ExprVal
+    String fraction = nanoStringVal.substring(nanoStringVal.indexOf(".") + 1);
+    fraction = StringUtils.rightPad(fraction, 9, "0");
+    long formattedNanos = Long.valueOf(fraction);
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1402]: Is it possible to avoid conversion `double -> string -> long`?
For example:
```java
value = value - Math.floor(value);
value *= 10^9;
value = value - Math.floor(value);
return Math.round(value); // to convert double to long
```

Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1402]: I've found it to be very difficult and unreliable no matter what method I use to keep it as a number. I've tried to do this math a few different ways, but there are always issues with imprecision when it comes to the step of multiplying by 10^9. I thought that it might be better to convert it to string even if there is potentially a performance hit because it guarantees a correct answer, but maybe it just requires an entirely different way to do the math than passing it into the `Duration.ofSeconds()` function. Any suggestions?

Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1402]:
```java
    var d = Double.valueOf(507116.921083338D);
    d = (d - Math.floor(d)) * 1_000_000_000;
    var r = Double.valueOf(Math.floor(d)).longValue();
```

Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java]: Altered implementation to use `BigDecimal` instead. https://github.com/Bit-Quill/opensearch-project-sql/pull/219/commits/3064a16f34af30b0908b10fd74dfc3769e5bc0cb

**Review by Yury-Fridlyand — APPROVED** [review]
